tests/Unit/NumericalAlgorithms/Interpolation/Test_LinearLeastSquares.cpp

```diff
@@ -41,6 +42,40 @@ void test_linear_least_squares(const std::array<double, Order + 1>& coeffs,
           my_approx(gsl::at(y_values, i)));
   }
 }
+
```

Comment:
Please add the following test:

- create a known set of N x_values in some range [x_lo, x_hi].  
- create K>2 (and K != N) unique sets of Order = 1 coeffs
- for each k in K  and n in N set y_values[k][n] = coeffs[k][1]*x_values[n] + coeffs[k][0]
- make sure `fit_coeffs` recovers `coeffs`
- for each k in K `interpolate` to a random point x in [x_lo, x_hi] and check that the returned value is coeffs[k][1]*x + coeffs[k][0]
- Fix this test as I don't think it will pass
- then instead of known x_values and coefficients use random values
- extend this test to higher Order where you never use K = Order + 1

Reply:
Hi Larry, can I ask how this test differs from that in `test_linear_least_squares_datavector`? From what I can tell,
N `x_values` is given by `datavecx`,
k>2 sets of coeffs of Order=3 are given by `datavector_coeffs`,
The values are set in line `55`,
The coeffs are checked in line `69`,
A point is interpolated to in line `71` and `73`.

The differences between what I have and the test you propose is that my test is Order 3 instead of Order 1, and that I interpolate to a value in `x_values` instead of a random value (though I also tested interpolating to a random value and the test passed) (I have also written a test of Order 1 which also passed)

Am I missing something key about how your proposed test is different? Sorry if I missed anything!

Reply:
It finds a bug in your code...
```
void test_linear_functions() {
  const DataVector x_values{1.0, 2.0, 3.0, 4.0, 5.0};
  const std::vector<std::array<double, 2>> coefficients{
      {{6.0, 7.0}}, {{8.0, 9.0}}, {{10.0, 11.0}}, {{12.0, 13.0}}};
  std::vector<DataVector> y_values{4, DataVector{5}};
  for (size_t i = 0; i < y_values.size(); ++i) {
    y_values[i] = x_values * coefficients[i][1] + coefficients[i][0];
  }
  intrp::LinearLeastSquares<1> lls(x_values.size());
  const std::vector<std::array<double, 2>> computed_coefficients =
      lls.fit_coefficients(x_values, y_values);
  CHECK_ITERABLE_APPROX(coefficients, computed_coefficients);
  for (size_t i = 0; i < y_values.size(); ++i) {
    double expected_y_value = 2.5 * coefficients[i][1] + coefficients[i][0];
    double computed_y_value = lls.interpolate(coefficients[i], 2.5);
    CHECK(expected_y_value == computed_y_value);
  }
}
```

Reply:
Hi Larry, I set up the coefficients this way:
```  const std::vector<std::array<double, 2>>& coeffs{{{1.0,2.0}},{{3.0,4.0}},{{5.0,6.0}}};
  const DataVector& x_values{0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0};
  std::vector<DataVector> y_values{};
  for (size_t i = 0; i < x_values.size(); i++) {
    //the size of a y_i DataVector is 3, there are 3 sets of coeffs (Number of Curves)
    DataVector y_i{coeffs.size(), 0.0};
    for (size_t j = 0; j < 2; j++) { // Loop over m and b
      for (size_t k = 0; k < coeffs.size(); k++) { // Loop over number of curves
        y_i[k] += gsl::at(coeffs[k], j) * pow(gsl::at(x_values, i), j);
      }
    }
    y_values.push_back(y_i);
  }
```

Comparing the two I realize now that the `DataVector` in `std::vector<DataVector> y_values` does not correspond to `DataVector` in `x_values` (though they are said to in the dox). The bug is that the data in `y_values` is transposed and my test was also transposed, so I didn't catch the disagreement with the documentation - I will fix this!

Reply:
Great!